index.js

```diff
@@ -1,39 +1,138 @@
-// declare some regular expressions and how to handle them
-const { expressions, BASE_SCC_URL } = require('./expressions.js');
+const {
+  BASE_SCC_URL,
+  CLASSIC_CATALOG_URL,
+} = process.env;
+
+
+const indexMappings = {
+  // X: '',
+  // Y: '',
+  a: '&search_scope=contributor',
+  t: '&search_scope=title',
+  s: '&search_scope=journal_title',
+  // h: '', // there is no genre search
+  // d: '', // this is the subject search, might want to do something different
+  i: '&search_scope=standard_number', // what to do with these?
+  c: '&search_scope=standard_number',
+}
+
+const getIndexMapping = index => indexMappings[index] || '';
+
+const recodeSearchQuery = query => query.split(/\+|\s/).join("%20");
+
+const getQueryFromParams = (url, query) => {
+  const getParam = param => (query[param] || [])[0];
+  const searchArg = getParam('searcharg');
+  const searchArgAlt = getParam('SEARCH');
+  let searchIndex;
+  let searchArgFromQueryParam;
+  for (let key in query) {
+    const value = getParam(key);
+    if (!value) {
+      const splitted = key.split('/');
+        if (splitted[1]) {
+          searchIndex = splitted[1][0];
+          searchArgFromQueryParam = splitted[1].slice(1);
+          break;
+        }
+        else if (splitted[0]) {
+          searchArgFromQueryParam = splitted[0];
+          break;
+        }
+    }
+  }
+  const q = searchArg || searchArgAlt || searchArgFromQueryParam || '';
+  const searchType = getParam('searchtype') || searchIndex;
+  return recodeSearchQuery(q + (searchType ? getIndexMapping(searchType) : ''));
+}
 
-// the main mapping method
+// declare some regular expressions
 
-const mapWebPacUrlToSCCURL = (url) => {
+const expressions = {
+  nothingReg: {
+    expr: /^\/$/,
+    handler: () => BASE_SCC_URL,
+  },
+  searchRegWith: {
+    expr: /\/search(~S\d*)?\/([a-zA-Z])(([^\/])+)/,
+    handler: match => `${BASE_SCC_URL}/search?q=${recodeSearchQuery(match[3])}${getIndexMapping(match[2])}`
+  },
+  searchRegWithout: {
+    expr: /\/search(~S\d*)?(\/([a-zA-Z]))?/,
+    handler: (match, query) => `${BASE_SCC_URL}/search?q=${getQueryFromParams(match[0], query)}${getIndexMapping(match[3])}`
+  },
+  patroninfoReg: {
+    expr: /\/patroninfo[^\/]\/(\d+)/,
+    handler: match => `${BASE_SCC_URL}?fakepatron=${match[1]}`
+  },
+  recordReg: {
+    expr: /\/record=(\w+)/,
+    handler: (match) => {
+      const bnum = match[1];
+      return `${BASE_SCC_URL}/bib/${bnum}`;
+    }
+  },
+};
+
+async function mapWebPacUrlToSCCURL(path, query) {
```

Comment:
Why is this async? I'm not seeing where it does async work.

Reply:
Good point, this is a relic of a previous implementation. Removed.